runtime/android/core/src/org/xwalk/core/XWalkInitializer.java

```diff
@@ -0,0 +1,132 @@
+    /**
+     * Interface used to initialize Crosswalk environment
+     */
+    public interface XWalkInitListener {
+        /**
+         * Runs on the UI thread to notify initialization is started
+         */
+        public void onXWalkInitStarted();
+        /**
+         * Runs on the UI thread to notify initialization is cancelled
+         */
+        public void onXWalkInitCancelled();
+        /**
+         * Runs on the UI thread to notify initialization is completed successfully
+         */
+        public void onXWalkInitCompleted();
+        /**
+         * Runs on the UI thread to notify initialization failed
+         */
+        public void onXWalkInitFailed();
+    }
+
+    private static XWalkInitializer sInstance;
+
+    /**
+     * Initializes Crosswalk environment asynchronously
+     * @param listener The {@link XWalkInitializer.XWalkInitListener} to use
+     * @param context The context that initialization is to run it
+     * @return False if is initializing, true otherwise
+     */
+    public static boolean initAsync(XWalkInitListener listener, Context context) {
+        if (sInstance != null) return false;
+        new XWalkInitializer(listener, context).start();
+        return true;
+    }
+
+    /**
+     * Attempts to cancel initialization
+     * @return False if initialization is not running or could not be cancelled, true otherwise
+     */
+    public static boolean cancelInit() {
+        if (sInstance == null) return false;
+        return sInstance.cancel();
+    }
+
+    private XWalkInitListener mListener;
+    private Context mContext;
+
+    private XWalkInitializer(XWalkInitListener listener, Context context) {
+        mListener = listener;
+        mContext = context;
+        XWalkLibraryLoader.prepareToInit();
+    }
+
+    @Override
+    public final void onDecompressionStarted() {
+        sInstance = this;
+        mListener.onXWalkInitStarted();
+    }
+
+    @Override
+    public final void onDecompressionCancelled() {
+        sInstance = null;
+        mListener.onXWalkInitCancelled();
+    }
+
+    @Override
+    public final void onDecompressionCompleted() {
+        int status = XWalkLibraryLoader.initXWalkLibrary(mContext);
+        if (status == XWalkLibraryInterface.STATUS_MATCH) {
+            XWalkLibraryLoader.startInitialization(this);
+        } else {
+            mListener.onXWalkInitFailed();
+        }
+    }
+
+    @Override
+    public final void onInitializationStarted() {
+        if (sInstance == null) {
+            sInstance = this;
+            mListener.onXWalkInitStarted();
+        }
+    }
+
+    @Override
+    public final void onInitializationCompleted() {
+        sInstance = null;
+        mListener.onXWalkInitCompleted();
+    }
+
+    private void start() {
+        if (XWalkLibraryLoader.startDecompression(this, mContext)) return;
+
+        final int status = XWalkLibraryLoader.initXWalkLibrary(mContext);
```

Comment:
startDecompression will execute an AsyncTask, is it too soon to call initXWalkLibrary after it?